### ISSUE TITLE
`ranges.h` blocks valid `formatter` tempaltes

### ISSUE THREAD

**Issue by fennewald** [fmtlib/fmt#3616]
In Certain Cases, adding `#include <fmt/ranges.h>` will break otherwise valid format specifiers. Here is an example with a custom `std::span` formatter:

```cpp
#include <fmt/format.h>

// Uncomment the next line to observe the issue
// #include <fmt/ranges.h>

#include <span>


template<typename T, std::size_t N>
struct fmt::formatter<std::span<const T, N>> {
    formatter<T> inner_formatter;

    constexpr auto parse(format_parse_context & ctx) {
        return inner_formatter.parse(ctx);
    }

    auto format(const std::span<const T, N> & spn, format_context & ctx) const {
        fmt::format_to(ctx.out(), "[");
        bool first = true;

        for (auto it : spn) {
            if (first) {
                first = false;
            } else {
                fmt::format_to(ctx.out(), ", ");
            }
            inner_formatter.format(it, ctx);
        }

        return fmt::format_to(ctx.out(), "]");
    }
};


int main() {
    constexpr int a[]{0, 1, 2, 3, 4, 5, 6, 7, 8};
    auto s = std::span{a};
    fmt::print("{}", s);
}
```

This compiles and functions as expected, but if the `#include <fmt/ranges.h>` is uncommented, it will fail at compile time.

[Here](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGIaa4AMngMmAByPgBGmMT%2BpAAOqAqETgwe3r5xicmOAsGhESzRsdJ2mA6pQgRMxATpPn6lmPa5DFU1BPnhUTFxCtW19ZmlA50h3UW9kgCUtqhexMjsHOYAzCHI3lgA1CarblQsBAD0/MQsTAQAdAh72CYaAIIPj8fH23IMaCxsgtsECEw21CqgI21o43%2BqG2qEiSmIADcYv9Ads8AoFF5MC83rszOsvlsgXsDkdjsRDMBMAobncXvT8Rsibt9gp4oY6U96U8CJgWPEDLySQQAJ7xRisIEAFVI2366BAIGSAC9MAB9MFhTmPfrELwObaHAiKs4XAi84gk%2BWKtkc/ZoBj9bYy7Za1b3d27ADsVie239hpIZotJKldzRDFCxDVpsuIdWvuefoDDv6mFU8WI2yYXiI23ZxCU1CDlzVBaUaodvNBeIAbNsHKppt7EwG29tiJgCAsGBGozGS%2BaYldy5gII3pntWwGTF6ACLcx5tnN52MEccCJ3WpXshgk1Ngl1u7B1uXxBiyteV2TpsHmesThubu8%2Bl7tgNGk2DjWoccEVRXPM66zHiZgmAArFYZhmJOCZvu%2B2yRKgniGnghZ3qsc7/Hq2JwVyybvmc2wQCu0KENsIBngwzaztOCEBngVDETQ6E0a%2BBH0e2LFOnsWFUGIShTvBnHelhzRKC2wkiR%2BRxfucpZEH%2BAFARAIHmGYsrqbBdGcbOC4cfRIT9muFpXGuECELKE5CQZM7zou9Gdt2xC9p%2BIBXopjaAbmqmadBEELtB2lSXp9L2XhSaRSEYIXCEqmSRxB7ppmEZgkwEFWOBC4%2BhospcLKGnbKssrSNs4GyrWsperKAAcoURcuubQgoLJYdutp7j66XhTpbmZtFEDqbRoXQbKCjBVy9lPBwsy0Jw4G8H4HBaKQqCcG41jWHK8yLMS%2BI8KQxrLTNswANYgOBuVzRwkiLZovBrRwvAKCAuVHVosxwLASDfPEdAxOQlC/f9sTAFwqwaVgCJ4EsABqeCYAA7gA8uKS0HTQtAWi9ECRPdpCRCENQipwB2E8wxAisjkTaOUR0Hd8vwEMjDC0CTx2kFgkReMAbhiLQL3cLwWAXEY4gc/gnYVEigsrem5S5ssB3Rc0%2BMQpEFKUx4WD4wQxB4CwpMzXwBjAAo8NI6jjBGzIggiGI7BSLb8hKGo%2BO6PlBhGCgm2WPoeBwvAEDMGw7lHKQSKxNwGgxydq3xK0gsPZH%2BtYC9sAh%2Bw8qRxHYhYj7ljWGYGjTLMZQVM4ECuEMfj5UE4yFMUejZCkAg183SStwwXSN70%2BXl607SDJ4DR6APlSjD3PSxP3ozt7PHRT5MM9lztSwSLN813Rzj3bKoNW1gAtLWkjbMAyDINs4NXGYxG4IQJB4qsXDTLw71x%2Bdl36Jwt2kEtK2PWeq9Q691S7fw4GYbeADOBv1AbMSOyRnCSCAA%3D%3D) is link to a reproduction in godbolt. Note that godbolt needed to be set to `c++20` to support `std::span`.

**Comment by fennewald** [comment]
Compiler error is as follows:
```text
In file included from /opt/compiler-explorer/gcc-13.2.0/include/c++/13.2.0/bits/stl_pair.h:60,
                 from /opt/compiler-explorer/gcc-13.2.0/include/c++/13.2.0/bits/stl_algobase.h:64,
                 from /opt/compiler-explorer/gcc-13.2.0/include/c++/13.2.0/bits/specfun.h:43,
                 from /opt/compiler-explorer/gcc-13.2.0/include/c++/13.2.0/cmath:3699,
                 from /opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/format.h:36,
                 from <source>:1:
/opt/compiler-explorer/gcc-13.2.0/include/c++/13.2.0/type_traits: In substitution of 'template<class _Tp, class ... _Args> using std::__is_constructible_impl = std::__bool_constant<__is_constructible(_Tp, _Args ...)> [with _Tp = fmt::v10::formatter<std::span<const int, 9>, char, void>; _Args = {}]':
/opt/compiler-explorer/gcc-13.2.0/include/c++/13.2.0/type_traits:1045:12:   required from 'struct std::is_constructible<fmt::v10::formatter<std::span<const int, 9>, char, void> >'
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:1441:51:   required from 'struct fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::formattable<const std::span<const int, 9>&, std::span<const int, 9> >'
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:1448:25:   required by substitution of 'template<class T, typename std::enable_if<(! fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::formattable<T, typename std::remove_cv<typename std::remove_reference<_Tp>::type>::type>::value), int>::type <anonymous> > constexpr fmt::v10::detail::unformattable fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::do_map(T&&) [with T = const std::span<const int, 9>&; typename std::enable_if<(! fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::formattable<T, typename std::remove_cv<typename std::remove_reference<_Tp>::type>::type>::value), int>::type <anonymous> = <missing>]'
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:1460:31:   required by substitution of 'template<class T, class U, typename std::enable_if<((((((std::is_class<U>::value || std::is_enum<U>::value) || std::is_union<U>::value) && (! fmt::v10::detail::is_string<U>::value)) && (! fmt::v10::is_char<U>::value)) && (! fmt::v10::detail::is_named_arg<T>::value)) && (! std::is_arithmetic<typename fmt::v10::detail::format_as_result<U>::type>::value)), int>::type <anonymous> > constexpr decltype (((fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >*)this)->fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::do_map(forward<T>(val))) fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::map(T&&) [with T = const std::span<const int, 9>&; U = std::span<const int, 9>; typename std::enable_if<((((((std::is_class<U>::value || std::is_enum<U>::value) || std::is_union<U>::value) && (! fmt::v10::detail::is_string<U>::value)) && (! fmt::v10::is_char<U>::value)) && (! fmt::v10::detail::is_named_arg<T>::value)) && (! std::is_arithmetic<typename fmt::v10::detail::format_as_result<U>::type>::value)), int>::type <anonymous> = 0]'
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:1476:53:   required by substitution of 'template<class T, class Context> using fmt::v10::detail::mapped_type_constant = fmt::v10::detail::type_constant<decltype (fmt::v10::detail::arg_mapper<Context>().map(declval<const T&>())), typename Context::char_type> [with T = std::span<const int, 9>; Context = fmt::v10::basic_format_context<fmt::v10::appender, char>]'
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:2621:66:   required from 'constexpr fmt::v10::detail::format_string_checker<Char, Args>::format_string_checker(fmt::v10::basic_string_view<Char>) [with Char = char; Args = {std::span<const int, 9>}]'
<source>:38:15:   required from here
<source>:38:15:   in 'constexpr' expansion of 'fmt::v10::basic_format_string<char, std::span<const int, 9>&>("{}")'
/opt/compiler-explorer/gcc-13.2.0/include/c++/13.2.0/type_traits:1040:25: error: ambiguous template instantiation for 'struct fmt::v10::formatter<std::span<const int, 9>, char, void>'
 1040 |       = __bool_constant<__is_constructible(_Tp, _Args...)>;
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from <source>:4:
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/ranges.h:560:8: note: candidates are: 'template<class R, class Char> struct fmt::v10::formatter<R, Char, typename std::enable_if<fmt::v10::conjunction<std::integral_constant<bool, (fmt::v10::range_format_kind<R, Char, void>::value != fmt::v10::range_format::disabled)>, fmt::v10::detail::is_formattable_delayed<R, Char> >::value, void>::type> [with R = std::span<const int, 9>; Char = char]'
  560 | struct formatter<
      |        ^~~~~~~~~~
  561 |     R, Char,
      |     ~~~~~~~~
  562 |     enable_if_t<conjunction<bool_constant<range_format_kind<R, Char>::value !=
      |     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  563 |                                           range_format::disabled>
      |                                           ~~~~~~~~~~~~~~~~~~~~~~~
  564 | // Workaround a bug in MSVC 2015 and earlier.
      | ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  565 | #if !FMT_MSC_VERSION || FMT_MSC_VERSION >= 1910
      | ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  566 |                             ,
      |                             ~
  567 |                             detail::is_formattable_delayed<R, Char>
      |                             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
  568 | #endif
      | ~~~~~~  
  569 |                             >::value>>
      |                             ~~~~~~~~~~
<source>:10:13: note:                 'template<class T, long unsigned int N> struct fmt::v10::formatter<std::span<const _Type, _ArrayExtent> > [with T = int; long unsigned int N = 9]'
   10 | struct fmt::formatter<std::span<const T, N>> {
      |             ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/compiler-explorer/gcc-13.2.0/include/c++/13.2.0/type_traits:1040:25: error: invalid use of incomplete type 'struct fmt::v10::formatter<std::span<const int, 9>, char, void>' [-fpermissive]
 1040 |       = __bool_constant<__is_constructible(_Tp, _Args...)>;
      |                         ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
In file included from /opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/format.h:49:
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:790:8: note: declaration of 'struct fmt::v10::formatter<std::span<const int, 9>, char, void>'
  790 | struct formatter {
      |        ^~~~~~~~~
/opt/compiler-explorer/gcc-13.2.0/include/c++/13.2.0/type_traits: In instantiation of 'struct std::is_constructible<fmt::v10::formatter<std::span<const int, 9>, char, void> >':
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:1441:51:   required from 'struct fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::formattable<const std::span<const int, 9>&, std::span<const int, 9> >'
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:1448:25:   required by substitution of 'template<class T, typename std::enable_if<(! fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::formattable<T, typename std::remove_cv<typename std::remove_reference<_Tp>::type>::type>::value), int>::type <anonymous> > constexpr fmt::v10::detail::unformattable fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::do_map(T&&) [with T = const std::span<const int, 9>&; typename std::enable_if<(! fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::formattable<T, typename std::remove_cv<typename std::remove_reference<_Tp>::type>::type>::value), int>::type <anonymous> = <missing>]'
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:1460:31:   required by substitution of 'template<class T, class U, typename std::enable_if<((((((std::is_class<U>::value || std::is_enum<U>::value) || std::is_union<U>::value) && (! fmt::v10::detail::is_string<U>::value)) && (! fmt::v10::is_char<U>::value)) && (! fmt::v10::detail::is_named_arg<T>::value)) && (! std::is_arithmetic<typename fmt::v10::detail::format_as_result<U>::type>::value)), int>::type <anonymous> > constexpr decltype (((fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >*)this)->fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::do_map(forward<T>(val))) fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::map(T&&) [with T = const std::span<const int, 9>&; U = std::span<const int, 9>; typename std::enable_if<((((((std::is_class<U>::value || std::is_enum<U>::value) || std::is_union<U>::value) && (! fmt::v10::detail::is_string<U>::value)) && (! fmt::v10::is_char<U>::value)) && (! fmt::v10::detail::is_named_arg<T>::value)) && (! std::is_arithmetic<typename fmt::v10::detail::format_as_result<U>::type>::value)), int>::type <anonymous> = 0]'
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:1476:53:   required by substitution of 'template<class T, class Context> using fmt::v10::detail::mapped_type_constant = fmt::v10::detail::type_constant<decltype (fmt::v10::detail::arg_mapper<Context>().map(declval<const T&>())), typename Context::char_type> [with T = std::span<const int, 9>; Context = fmt::v10::basic_format_context<fmt::v10::appender, char>]'
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:2621:66:   required from 'constexpr fmt::v10::detail::format_string_checker<Char, Args>::format_string_checker(fmt::v10::basic_string_view<Char>) [with Char = char; Args = {std::span<const int, 9>}]'
<source>:38:15:   required from here
<source>:38:15:   in 'constexpr' expansion of 'fmt::v10::basic_format_string<char, std::span<const int, 9>&>("{}")'
/opt/compiler-explorer/gcc-13.2.0/include/c++/13.2.0/type_traits:1048:52: error: static assertion failed: template argument must be a complete class or an unbounded array
 1048 |       static_assert(std::__is_complete_or_unbounded(__type_identity<_Tp>{}),
      |                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~
/opt/compiler-explorer/gcc-13.2.0/include/c++/13.2.0/type_traits:1048:52: note: 'std::__is_complete_or_unbounded<__type_identity<fmt::v10::formatter<span<const int, 9>, char, void> > >((std::__type_identity<fmt::v10::formatter<std::span<const int, 9>, char, void> >(), std::__type_identity<fmt::v10::formatter<std::span<const int, 9>, char, void> >()))' evaluates to false
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h: In instantiation of 'struct fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::formattable<const std::span<const int, 9>&, std::span<const int, 9> >':
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:1448:25:   required by substitution of 'template<class T, typename std::enable_if<(! fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::formattable<T, typename std::remove_cv<typename std::remove_reference<_Tp>::type>::type>::value), int>::type <anonymous> > constexpr fmt::v10::detail::unformattable fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::do_map(T&&) [with T = const std::span<const int, 9>&; typename std::enable_if<(! fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::formattable<T, typename std::remove_cv<typename std::remove_reference<_Tp>::type>::type>::value), int>::type <anonymous> = <missing>]'
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:1460:31:   required by substitution of 'template<class T, class U, typename std::enable_if<((((((std::is_class<U>::value || std::is_enum<U>::value) || std::is_union<U>::value) && (! fmt::v10::detail::is_string<U>::value)) && (! fmt::v10::is_char<U>::value)) && (! fmt::v10::detail::is_named_arg<T>::value)) && (! std::is_arithmetic<typename fmt::v10::detail::format_as_result<U>::type>::value)), int>::type <anonymous> > constexpr decltype (((fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >*)this)->fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::do_map(forward<T>(val))) fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::map(T&&) [with T = const std::span<const int, 9>&; U = std::span<const int, 9>; typename std::enable_if<((((((std::is_class<U>::value || std::is_enum<U>::value) || std::is_union<U>::value) && (! fmt::v10::detail::is_string<U>::value)) && (! fmt::v10::is_char<U>::value)) && (! fmt::v10::detail::is_named_arg<T>::value)) && (! std::is_arithmetic<typename fmt::v10::detail::format_as_result<U>::type>::value)), int>::type <anonymous> = 0]'
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:1476:53:   required by substitution of 'template<class T, class Context> using fmt::v10::detail::mapped_type_constant = fmt::v10::detail::type_constant<decltype (fmt::v10::detail::arg_mapper<Context>().map(declval<const T&>())), typename Context::char_type> [with T = std::span<const int, 9>; Context = fmt::v10::basic_format_context<fmt::v10::appender, char>]'
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:2621:66:   required from 'constexpr fmt::v10::detail::format_string_checker<Char, Args>::format_string_checker(fmt::v10::basic_string_view<Char>) [with Char = char; Args = {std::span<const int, 9>}]'
<source>:38:15:   required from here
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:1441:51: error: 'value' is not a member of 'std::is_constructible<fmt::v10::formatter<std::span<const int, 9>, char, void> >'
 1441 |                       (has_formatter<U, Context>::value &&
      |                                                   ^~~~~
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h: In instantiation of 'constexpr decltype (ctx.begin()) fmt::v10::detail::parse_format_specs(ParseContext&) [with T = std::span<const int, 9>; ParseContext = compile_parse_context<char>; decltype (ctx.begin()) = const char*]':
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:2620:22:   required from 'constexpr fmt::v10::detail::format_string_checker<Char, Args>::format_string_checker(fmt::v10::basic_string_view<Char>) [with Char = char; Args = {std::span<const int, 9>}]'
<source>:38:15:   required from here
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:2561:10: error: invalid use of incomplete type 'struct fmt::v10::formatter<std::span<const int, 9>, char, void>'
 2561 |   return formatter<mapped_type, char_type>().parse(ctx);
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:790:8: note: declaration of 'struct fmt::v10::formatter<std::span<const int, 9>, char, void>'
  790 | struct formatter {
      |        ^~~~~~~~~
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h: In instantiation of 'struct fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::formattable<std::span<const int, 9>&, std::span<const int, 9> >':
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:1448:25:   required by substitution of 'template<class T, typename std::enable_if<(! fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::formattable<T, typename std::remove_cv<typename std::remove_reference<_Tp>::type>::type>::value), int>::type <anonymous> > constexpr fmt::v10::detail::unformattable fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::do_map(T&&) [with T = std::span<const int, 9>&; typename std::enable_if<(! fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::formattable<T, typename std::remove_cv<typename std::remove_reference<_Tp>::type>::type>::value), int>::type <anonymous> = <missing>]'
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:1460:31:   required by substitution of 'template<class T, class U, typename std::enable_if<((((((std::is_class<U>::value || std::is_enum<U>::value) || std::is_union<U>::value) && (! fmt::v10::detail::is_string<U>::value)) && (! fmt::v10::is_char<U>::value)) && (! fmt::v10::detail::is_named_arg<T>::value)) && (! std::is_arithmetic<typename fmt::v10::detail::format_as_result<U>::type>::value)), int>::type <anonymous> > constexpr decltype (((fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >*)this)->fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::do_map(forward<T>(val))) fmt::v10::detail::arg_mapper<fmt::v10::basic_format_context<fmt::v10::appender, char> >::map(T&&) [with T = std::span<const int, 9>&; U = std::span<const int, 9>; typename std::enable_if<((((((std::is_class<U>::value || std::is_enum<U>::value) || std::is_union<U>::value) && (! fmt::v10::detail::is_string<U>::value)) && (! fmt::v10::is_char<U>::value)) && (! fmt::v10::detail::is_named_arg<T>::value)) && (! std::is_arithmetic<typename fmt::v10::detail::format_as_result<U>::type>::value)), int>::type <anonymous> = 0]'
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:1674:41:   required from 'constexpr fmt::v10::detail::value<Context> fmt::v10::detail::make_value(T&&) [with Context = fmt::v10::basic_format_context<fmt::v10::appender, char>; T = std::span<const int, 9>&]'
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:1711:29:   required from 'constexpr fmt::v10::detail::value<Context> fmt::v10::detail::make_arg(T&&) [with bool IS_PACKED = true; Context = fmt::v10::basic_format_context<fmt::v10::appender, char>; type <anonymous> = fmt::v10::detail::type::custom_type; T = std::span<const int, 9>&; typename std::enable_if<IS_PACKED, int>::type <anonymous> = 0]'
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:1827:77:   required from 'constexpr fmt::v10::format_arg_store<Context, Args>::format_arg_store(T&& ...) [with T = {std::span<const int, 9>&}; Context = fmt::v10::basic_format_context<fmt::v10::appender, char>; Args = {std::span<const int, 9>}]'
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:1844:31:   required from 'constexpr fmt::v10::format_arg_store<Context, typename std::remove_cv<typename std::remove_reference<T>::type>::type ...> fmt::v10::make_format_args(T&& ...) [with Context = basic_format_context<appender, char>; T = {std::span<const int, 9>&}]'
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:2904:44:   required from 'void fmt::v10::print(format_string<T ...>, T&& ...) [with T = {std::span<const int, 9>&}; format_string<T ...> = basic_format_string<char, std::span<const int, 9>&>]'
<source>:38:15:   required from here
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:1441:51: error: 'value' is not a member of 'std::is_constructible<fmt::v10::formatter<std::span<const int, 9>, char, void> >'
 1441 |                       (has_formatter<U, Context>::value &&
      |                                                   ^~~~~
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h: In instantiation of 'constexpr fmt::v10::detail::value<Context> fmt::v10::detail::make_value(T&&) [with Context = fmt::v10::basic_format_context<fmt::v10::appender, char>; T = std::span<const int, 9>&]':
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:1711:29:   required from 'constexpr fmt::v10::detail::value<Context> fmt::v10::detail::make_arg(T&&) [with bool IS_PACKED = true; Context = fmt::v10::basic_format_context<fmt::v10::appender, char>; type <anonymous> = fmt::v10::detail::type::custom_type; T = std::span<const int, 9>&; typename std::enable_if<IS_PACKED, int>::type <anonymous> = 0]'
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:1827:77:   required from 'constexpr fmt::v10::format_arg_store<Context, Args>::format_arg_store(T&& ...) [with T = {std::span<const int, 9>&}; Context = fmt::v10::basic_format_context<fmt::v10::appender, char>; Args = {std::span<const int, 9>}]'
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:1844:31:   required from 'constexpr fmt::v10::format_arg_store<Context, typename std::remove_cv<typename std::remove_reference<T>::type>::type ...> fmt::v10::make_format_args(T&& ...) [with Context = basic_format_context<appender, char>; T = {std::span<const int, 9>&}]'
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:2904:44:   required from 'void fmt::v10::print(format_string<T ...>, T&& ...) [with T = {std::span<const int, 9>&}; format_string<T ...> = basic_format_string<char, std::span<const int, 9>&>]'
<source>:38:15:   required from here
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:1691:7: error: static assertion failed: Cannot format an argument. To make type T formattable provide a formatter<T> specialization: https://fmt.dev/latest/api.html#udt
 1691 |       formattable,
      |       ^~~~~~~~~~~
/opt/compiler-explorer/libs/fmt/10.0.0/include/fmt/core.h:1691:7: note: 'formattable' evaluates to false
Compiler returned: 1
```

**Comment by vitaut** [comment]
The error is expected because you are defining a conflicting formatter for a range. Either use the standard range formatter or use a wrapper type.

**Comment by fennewald** [comment]
Ahh, I had failed to format a `std::span` due to a different mistake of my own. Thank you for your patience!

**Comment by petersteneteg** [comment]
I also ran into this issue. This is a major problem! There is no way of specifying a custom formatter for a template type that has a begin and end unless you can provide a full specialization. And like in the example above there is no way to specialize span for every value of N... And for library types it is very hard to ensure that fmt/ranges.h is never seen at the same time, other types might need if for formatting... 

basically I have to either remove begin() and end() from my type or ban any use of fmt/ranges.h

**Comment by petersteneteg** [comment]
going through more issues, I found this workaround:

`
template <typename T>
struct fmt::is_range<blaaah<T>, char> : std::false_type {};
`

Would have liked to se that in the docs...
